runtime/vm/extendedMessageNPE.cpp

```diff
@@ -291,43 +383,85 @@ getCompleteNPEMessage(J9VMThread *vmThread, U_8 *bcCurrentPtr, J9ROMClass *romCl
 		default:
 			Trc_VM_GetCompleteNPEMessage_Unreachable(vmThread, bcCurrent);
 		}
-		if ((bcCurrent >= JBiaload) && (bcCurrent <= JBsaload)) {
-			msgTemplate = "Cannot load from %s array";
+		if (NULL == npeCauseMsg) {
+			if ((bcCurrent >= JBiaload) && (bcCurrent <= JBsaload)) {
+				msgTemplate = "Cannot load from %s array";
+			} else {
+				msgTemplate = "Cannot store to %s array";
+			}
+			npeMsg = getMsgWithAllocation(vmThread, msgTemplate, elementType);
 		} else {
-			msgTemplate = "Cannot store to %s array";
+			if ((bcCurrent >= JBiaload) && (bcCurrent <= JBsaload)) {
+				if (isMethodFlag) {
+					msgTemplate = "Cannot load from %s array because the return value of \"%s\" is null";
+				} else {
+					msgTemplate = "Cannot load from %s array because \"%s\" is null";
+				}
+			} else {
+				if (isMethodFlag) {
+					msgTemplate = "Cannot store to %s array because the return value of \"%s\" is null";
+				} else {
+					msgTemplate = "Cannot store to %s array because \"%s\" is null";
+				}
+			}
+			npeMsg = getMsgWithAllocation(vmThread, msgTemplate, elementType, npeCauseMsg);
 		}
-		msgLen = j9str_printf(PORTLIB, NULL, 0, msgTemplate, elementType);
-		/* msg NULL check omitted since str_printf accepts NULL (as above) */
-		npeMsg = (char *)j9mem_allocate_memory(msgLen + 1, OMRMEM_CATEGORY_VM);
-		j9str_printf(PORTLIB, npeMsg, msgLen, msgTemplate, elementType);
 	} else {
 		switch (bcCurrent) {
 		case JBarraylength: {
-			const char *VM_NPE_ARRAYLENGTH = "Cannot read the array length";
-			msgLen = j9str_printf(PORTLIB, NULL, 0, VM_NPE_ARRAYLENGTH);
-			npeMsg = (char *)j9mem_allocate_memory(msgLen + 1, OMRMEM_CATEGORY_VM);
-			j9str_printf(PORTLIB, npeMsg, msgLen, VM_NPE_ARRAYLENGTH);
+			if (NULL == npeCauseMsg) {
+				npeMsg = getMsgWithAllocation(vmThread, msgTemplate, "Cannot read the array length");
+			} else {
+				const char *VM_NPE_ARRAYLENGTH = NULL;
```

Comment:
I dislike all caps variable names - using `msgTemplate` like above would be better.

Reply:
Fixed. Please have another look.